code/go/0chain.net/smartcontract/storagesc/allocation_test.go

```diff
@@ -230,6 +232,219 @@ func TestSelectBlobbers(t *testing.T) {
+		}
+
+		alloc := &StorageAllocation{
+			ID:              mockAllocationID,
+			Owner:           mockOwner,
+			BlobberDetails:  blobberAllocation,
+			BlobberMap:      blobberMap,
+			Name:            mockAllocationName,
+			Size:            confMinAllocSize,
+			Expiration:      mockAllocationExpiry,
+			ReadPriceRange:  PriceRange{mockMinPrice, mockMaxPrice},
+			WritePriceRange: PriceRange{mockMinPrice, mockMaxPrice},
+			DataShards:      arg.dataShards,
+			ParityShards:    arg.parityShards,
+		}
+
+		if len(arg.addBlobberID) > 0 {
+			blobber := &StorageNode{
+				ID: arg.addBlobberID,
+				Terms: Terms{
+					MaxOfferDuration: mockMaxOffDuration,
+					ReadPrice:        mockReadPrice,
+					WritePrice:       mockWritePrice,
+				},
+				Capacity:        mockBlobberCapacity,
+				LastHealthCheck: now,
+			}
+			balances.On("GetTrieNode", blobber.GetKey(sc.ID), mock.MatchedBy(func(s *StorageNode) bool {
+				*s = *blobber
+				return true
+			})).Return(nil).Once()
+
+			sp := stakePool{
+				StakePool: stakepool.StakePool{
+					Pools: map[string]*stakepool.DelegatePool{
+						mockPoolId: {},
+					},
+				},
+			}
+			sp.Pools[mockPoolId].Balance = mockState
+			balances.On("GetTrieNode", stakePoolKey(sc.ID, arg.addBlobberID),
+				mock.MatchedBy(func(s *stakePool) bool {
+					*s = sp
+					return true
+				})).Return(nil).Twice()
+		}
+
+		if len(arg.removeBlobberID) > 0 {
+			blobber := &StorageNode{
+				ID:       arg.removeBlobberID,
+				Capacity: mockBlobberCapacity,
+				Terms: Terms{
+					MaxOfferDuration: mockMaxOffDuration,
+					ReadPrice:        mockReadPrice,
+					WritePrice:       mockWritePrice,
+				},
+				LastHealthCheck: now,
+			}
+
+			balances.On(
+				"InsertTrieNode",
+				blobber.GetKey(sc.ID),
+				mock.Anything,
+			).Return("", nil).Once()
+
+			balances.On(
+				"EmitEvent",
+				event.TypeStats,
+				event.TagUpdateBlobber,
+				blobber.ID,
+				mock.Anything,
+			).Return("", nil).Once()
+		}
+
+		return blobbers, arg.addBlobberID, arg.removeBlobberID, &sc, alloc, now, balances
+
+	}
+
+	testCases := []struct {
+		name string
+		args args
+		want want
+	}{
+		{
+			name: "test_blobber_remove_doesnt_exist",
+			args: args{
+				numBlobbers:     6,
+				addBlobberID:    "add_blobber_id",
+				removeBlobberID: "blobber_non_existent",
+				dataShards:      5,
+			},
+			want: want{
+				err:    true,
+				errMsg: "cannot find blobber blobber_non_existent in allocation",
+			}},
+		{
+			name: "test_blobber_add_remove_blobber",
+			args: args{
+				numBlobbers:     6,
+				addBlobberID:    "add_blobber_id",
+				removeBlobberID: "blobber_1",
+				dataShards:      5,
+			},
+			want: want{
+				err: false,
+			}},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			blobbers, addID, removeID, sc, sa, now, balances := setup(tt.args)
+			_, err := sa.changeBlobbers(blobbers, addID, removeID, sc, now, balances)
+			require.EqualValues(t, tt.want.err, err != nil)
```

Comment:
Please assert the state as well, if we removed a blobber, please check and assert the blobber is removed, and related state were updated correctly. Checking the error message is literally useless, we need to check if the state is changed as we expected.